src/recommendation/src/main/scala/RecommendationIndexer.scala

```diff
@@ -0,0 +1,160 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.attribute.NominalAttribute
+import org.apache.spark.ml.feature.{StringIndexer, StringIndexerModel}
+import org.apache.spark.ml.param.{Param, ParamMap, Params, TransformerParam}
+import org.apache.spark.ml.util._
+import org.apache.spark.ml.{Estimator, Model, Transformer}
+import org.apache.spark.sql.expressions.UserDefinedFunction
+import org.apache.spark.sql.functions.udf
+import org.apache.spark.sql.types.{NumericType, StringType, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+class RecommendationIndexer(override val uid: String)
+  extends Estimator[RecommendationIndexerModel] with RecommendationIndexerBase {
+
+  def this() = this(Identifiable.randomUID("RecommendationIndexer"))
+
+  override def fit(dataset: Dataset[_]): RecommendationIndexerModel = {
+    val userIndexModel: StringIndexerModel = new StringIndexer()
+      .setInputCol(getUserInputCol)
+      .setOutputCol(getUserOutputCol)
+      .fit(dataset)
+
+    val itemIndexModel: StringIndexerModel = new StringIndexer()
+      .setInputCol(getItemInputCol)
+      .setOutputCol(getItemOutputCol)
+      .fit(dataset)
+
+    new RecommendationIndexerModel(uid)
+      .setParent(this)
+      .setUserIndexModel(userIndexModel)
+      .setItemIndexModel(itemIndexModel)
+      .setUserInputCol(getUserInputCol)
+      .setUserOutputCol(getUserOutputCol)
+      .setItemInputCol(getItemInputCol)
+      .setItemOutputCol(getItemOutputCol)
+  }
+
+  override def copy(extra: ParamMap): Estimator[RecommendationIndexerModel] = defaultCopy(extra)
+
+}
+
+object RecommendationIndexer extends ComplexParamsReadable[RecommendationIndexer]
+
+class RecommendationIndexerModel(override val uid: String) extends Model[RecommendationIndexerModel] with
+  RecommendationIndexerBase with Wrappable {
+  override def copy(extra: ParamMap): RecommendationIndexerModel = defaultCopy(extra)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    getItemIndexModel.transform(getUserIndexModel.transform(dataset))
+  }
+
+  def this() = this(Identifiable.randomUID("RecommendationIndexerModel"))
+
+  val userIndexModel = new TransformerParam(this, "userIndexModel", "userIndexModel", { x: Transformer => true })
+
+  def setUserIndexModel(m: Model[_]): this.type = set(userIndexModel, m)
```

Comment:
Make this the right type. If you want it to only take in a specific model nows your chance to enforce constraints